agera/src/main/java/com/google/android/agera/Functions.java

```diff
@@ -110,5 +133,47 @@ public T apply(@NonNull F from) {
     }
   }
 
+  private static final class ParallelMapFunction<F, T> implements Function<List<F>, List<T>> {
+    @NonNull
+    private final Function<F, T> function;
+    @NonNull
+    private final ExecutorService executorService;
+
+    ParallelMapFunction(@NonNull final ExecutorService executorService,
+        @NonNull final Function<F, T> function) {
+      this.executorService = checkNotNull(executorService);
+      this.function = checkNotNull(function);
+    }
+
+    @NonNull
+    @Override
+    @SuppressWarnings("unchecked")
+    public List<T> apply(@NonNull final List<F> input) {
+      if (input.isEmpty()) {
+        return emptyList();
+      }
+      final List<T> result = new ArrayList(input.size());
+      final List<Future<T>> futureResult = new ArrayList(input.size());
+      for (final F item : input) {
+        futureResult.add(executorService.submit(new Callable<T>() {
+          @Override
+          public T call() throws Exception {
+            return function.apply(item);
+          }
+        }));
+      }
+      final int size = futureResult.size();
+      for (int i = 0; i < size; i++) {
+        final Future<T> item = futureResult.get(i);
+        try {
+          result.add(item.get());
+        } catch (final ExecutionException | InterruptedException e) {
+          result.add(function.apply(input.get(i)));
```

Comment:
Or even List<Result<T>>? :P I'd allow for some execution to fail, and some to succeed. then again, it'd mean we might get List<Result<Result<Stuff>> if the type is Result<Stuff>. I'm not too happy with either solution. we should think about this some more...